### ISSUE TITLE
entity_url returns instance of Drupal\\Core\Url

### ISSUE THREAD

**Issue by jordonedavidson** [drupal-graphql/graphql#858]
I'm working to understand the complexities of this service, and I seem to keep running into major roadblocks on what should, IMHO, be pretty straightforward tasks.

The docs for the built-in producer `entity_url` says "Returns the entity's url." In pretty much every other case of a simple producer like that (`entity_label`, `entity_language`, et al) a simple string value or null is returned. 

When setting up my schema like this:
```
type Article {
  id: Int!
  title: String!
  author: String
  url: String
  body: String
  summary: String
}
```

with this resolver:
```
    $registry->addFieldResolver(
      'Article',
      'url',
      $builder->produce('entity_url')
        ->map('entity', $builder->fromParent())
    );
```

I get the error: `"Expected a value of type \"String\" but received: instance of Drupal\\Core\\Url"`

I don't understand a) why the ->toURL() function would return a class (although further rabbit holing into the bowels of Drupal may yield that answer.) and b) How to convert this class into the actual url string I desire.

Documentation is hard, and this whole GraphQL/Doctrine/Drupal mash up is quite complex. We definitely need more, well commented, examples that show exactly what is happening and, more importantly, what needs to happen for common use cases.

**Comment by jordonedavidson** [comment]
The answer seems to lie with the producer `url_path` as it i looking for an instance of Drupal\Core\Url. Unfortunately I can't grok how to string them together in the resolver

**Comment by DiegoPino** [comment]
as you noticed `::toURL()` returns an URL Object not a String
 see https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Entity%21Entity.php/function/Entity%3A%3AtoUrl/8.7.x

to return the corresponding string representation you need to invoke ::toString() method at the end
I guess (can be wrong) in this case $builder->produce('entity_url')->map('entity', $builder->fromParent()->toString(). Reason for that (i guess in this context) is that there are many possible representations of an URL in string format (relative, absolute, canonical, etc). See the whole URL class definition here to experiment with other options. https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Url.php/class/Url/8.7.x

Probably the issue here is documentation but also the fact that in Drupal 8 slang an URL refers always to the URL Object. Also, on the positive side this makes that producer way more flexible!

**Comment by jordonedavidson** [comment]
> to return the corresponding string representation you need to invoke ::toString() method at the end
> I guess in this case $builder->produce('entity_url')->toString(). 

Unfortunately adding the ->toString() option to the ResolverInterface parameter breaks it. (Which makes good sense.) 

I think the `url_path` resolver is the key, I just can't figure out the correct syntax to pass the result of `entity_url` to it.

**Comment by DiegoPino** [comment]
@jordonedavidson maybe by using the composable schema methods (using the 'entity_url' processor first and then the 'url_path') ?

See

https://github.com/drupal-graphql/graphql/blob/8.x-4.x/doc/advanced/composable-schemas.md

**Comment by jordonedavidson** [comment]
@DiegoPino Yep!

```
$registry->addFieldResolver(
      'Article',
      'url',
      $builder->compose(
        $builder->produce('entity_url')
          ->map('entity', $builder->fromParent()),
        $builder->produce('url_path')
          ->map('url', $builder->fromParent())
      )
    );
```
Did the trick. 

It's not super clear from the docs as to exactly how this is working, but it is.

For instance: I just guessed that I needed to map 'url' and not 'path' or 'value' in the url_path line, and it worked. No idea why, or where I could have discovered this in the documentation had I not guessed correctly.

Anyway, I'm closing this and leaving it here for the next Drupal/GraphQL newbie who comes along.

**Comment by fubhy** [comment]
@jordonedavidson Yes, the lack of documentation is definitely unfortunate. Documentation is hard. We are always looking for help especially for documentation. If you'd like to get involved, feel free to swing by our Slack channel on the Drupal Slack community. Also of course for any other questions around building a schema with 4.x - We are quite active there.

One important and simple hint regarding the way resolvers work: At their core, each resolver (anything that is returned from any of the `$builder` methods (`->compose`, `->produce`, etc.) is a callable class instance that follows this simple interface: https://github.com/drupal-graphql/graphql/blob/8.x-4.x/src/GraphQL/Resolver/ResolverInterface.php

Essentially, every piece of code that runs in our schemas to compute data is nothing but a class that has a `resolve()` method. Now, there are simple implementations that allow you to pass in a closure/callback (https://github.com/drupal-graphql/graphql/blob/8.x-4.x/src/GraphQL/Resolver/Callback.php) like so:

```
$builder->callback(function ($parent, $args) {
  return "Hello world!";
});
```

and there are more complex ones like `$builder->compose()` that allow you to chain together multiple other resolvers (e.g. instances of DataProducers (`$builder->produce()`).

Since we strive for max. flexibility to allow full customization when building the schema that most suites your consuming applications use-case, we try to provide an (exhaustive) set of small DataProducers that allow you to compose them together in the way that you need them for your schema instead of providing a huge amount of specialized, non-reusable monolithic DataProducers. 

TLDR: The stuff that you pass to the `->addFieldResolver()` method is just a simple class instance with a `->resolve()` method and we provide some helpers (through ResolverBuilder) that allow you compose your resolvers together from DataProducers (plugins) and through other means. But at the end of the day, everything is just a callable.
